### ISSUE TITLE
Reconstruction of interpolated function in the whole domain with GPU?

### ISSUE THREAD

**Issue by RaimelMedina** [tensor4all/TensorCrossInterpolation.jl#36]
Hi again,

I'm using TCI for compressing a very expensive function (an expectation value of a parameterized N-qubit state in my case) in a specific grid. In my case, the parameters are angles and are in the (-\pi/4, \pi/4), and the grid dimension is dim=2, 4, 6, 8. With the tci at hand I then reconstruct the function in the whole grid (I need it for some data analysis). 

The above is faster than calling the original function but I was wondering if it is possible to speed up the reconstruction with the help of a GPU given the parallel nature of the problem. 

I would appreciate any suggestions/advice you could give me,

Thanks!

**Comment by shinaoka** [comment]
You could convert a `TCI.TensorCI2` object to a `TCI.TensorTrain`, and then to an `ITensors.MPS` object, by using `TCIITensorConversion.jl`.

Using `ITensors.jl`, you could reconstruct a full tensor by simply doing `reduce(*, M)` where `M` is an MPS.
This contract all tensor cores  in `M` by matrix multiplications.
This contraction could be accelerated by `ITensorGPU.jl`, or could be fast even with a CPU.

**Comment by shinaoka** [comment]
@RaimelMedina 
Alternatively, we can update `TensorCrossInterpolation.jl` to version 0.9.11, which will be released in a couple of hours.
Then, you could reconstruct the full tensor by 

```julia
tci, ranks, errors = TCI.crossinterpolate1(ComplexF64, g, localdims; tolerance=tolerance)
tt = TCI.TensorTrain(tci)
TCI.fulltensor(tt)
```

From v0.9.11, this function uses matrix multiplications accelerated by lapack, which should be much faster than evaluating the tensor train for all indices independently.

If you want to run it on a GPU, you can hack the code here:
https://github.com/tensor4all/TensorCrossInterpolation.jl/blob/main/src/tensortrain.jl#L256

**Comment by RaimelMedina** [comment]
@shinaoka 
Thanks a lot!!! I recently tested the ITensor approach and I have to say is very fast in comparison to what I had before 😅, which was evaluating the tensor train for all indices independently (in parallel). Like 15 minutes vs 4.5 seconds for a grid of 20^6 points. I will try the other suggestion you gave me as well now. 

Don't know if I should open a new issue for this but: Should one expect a considerable drop in the convergence speed in constructing the tci when going from Float64-> Float32? The reason I'm asking is because on the GPU support (Metal only supports single precision :( )

**Comment by shinaoka** [comment]
Good news!

We recommend to use `crossinterpolate2(Float64, ...)` for numerical stability.

But, you can evaluate your function using single precision and convert a Float32 to a Float64 as follows:

```julia
fdouble(x) = Float64(fsingle(x)) # fsingle is your function returning a Float32. 
crossinterpolate2(Float64, fdouble, ....)
```

This does not affect the convergence speed of TCI if your tolerance for TCI is not too low.

**Comment by RaimelMedina** [comment]
Hi again,

I just tested the new `fulltensor` function. It works nicely! It is slower than constructing and contracting the MPS in ITensor though, but it works pretty nicely. I wonder where the speed in the ITensor approach comes from... 
I will try as well to come up with a GPU version of the code.

Thanks again

Benchmark on a grid of 20^6 points

- `fulltensor`
```
@time enew_app =  TCI.fulltensor(TensorTrain(tci_1));
502.495129 seconds (896.63 M allocations: 2.214 TiB, 19.75% gc time, 0.03% compilation time)
```
- `ITensor.MPS`
```
julia> @time egrid_new = reduce(*, M);
  4.540203 seconds (20.46 M allocations: 1.878 GiB, 8.00% gc time, 97.83% compilation time)

julia> @time egrid_new = reduce(*, M);
  0.135744 seconds (159 allocations: 575.703 MiB, 9.15% gc time)
 ```

**Comment by shinaoka** [comment]
Could you confirm that  the version of your `TensorCrossInterpolation.jl` is 0.9.12? `fulltensor` in 0.9.11 is slow.

**Comment by RaimelMedina** [comment]
Oh ok ok, I didn't update it in my environment, my bad :|
Now it is way better 😀 Apologies for the mistake!
```
julia> @time egrid_full = TCI.fulltensor(TCI.TensorTrain(tci_1));
  0.304302 seconds (48 allocations: 575.689 MiB, 75.44% gc time)

julia> @time egrid_full = TCI.fulltensor(TCI.TensorTrain(tci_1));
  0.081287 seconds (48 allocations: 575.689 MiB)
```
